### ISSUE TITLE
S2583 + S2589: False positive in forEach loop

### ISSUE THREAD

**Issue by Villae** [SonarSource/SonarJS#983]
SonarJS version: SonarJS 4.0.0.5862 (used in the VS Code sonarlint extension v1.2.0)
Also tested with SonarJS 4.1.0.6085 version. (Manually upgraded the jar in VS Code extension folder)

Rule key: S2583 + S2589 (I did't get the S2583 error in my real code, but it reports error in this minimal example, weird)

**Reproducer**
```
const tooltipData = []
const absences = [{ absenceCode: 'randomcode' }] // Just placeholder for this example

absences.forEach(absence => {
  const absenceCode = absence.absenceCode
  let toolTipText = ''
  let absenceClass = ''

  if (absenceCode === '1') {
    absenceClass = 'class1'
    toolTipText = 'random1'
  } else if (absenceCode === '2') {
    absenceClass = 'class2'
    toolTipText = 'random2'
  }

  if (toolTipText || absenceClass) { // False positive #1: Change this condition so that it does not always evaluate to "true"; some subsequent code is never executed. (javascript:S2583)
    tooltipData.push({
      classes: absenceClass // False positive #2 here: Refactor this code so that this expression does not always evaluate to false. (javascript:S2589)
        ? [absenceClass, 'extra-class']
        : [],
      tooltip: toolTipText
    })
  }
})

return tooltipData
```

**Expected behavior**
No errors? :) 

**Note**
I can reproduce this only in the arr.forEach loop. Without loop or in for () loop, the error does not occur.

**Comment by andrea-guarino-sonarsource** [comment]
Hi @Villae,
Although not very easy to spot, the issues you got are both relevant.

Fist of all, I believe the two comments about *false positive* are reversed: "False positive 1" refers to the ternary operator condition while "False positive 2" refers to `absenceClass` in `toolTipText || absenceClass` expression.

### **Issue on rule S2589**

```javascript
 if (toolTipText || absenceClass) {  // Refactor this code so that this expression does not always evaluate to false. (javascript:S2589)
```
Notice that `tooltipText` and `absenceClass` could be only either both empty string (*falsy* value) or both non-empty string (*truthy* value). In first case, when both are empty string, `absenceClass` will evaluate to *false*. In second case, when both are not empty, `absenceClass` won't be evaluated at all because of [short-circuit evaluation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Logical_Operators).

### **Issue on rule S2583**
```javascript
if (toolTipText || absenceClass) {
  tooltipData.push({
     absenceClass // Change this condition so that it does not always evaluate to "true"; some subsequent code is never executed. (javascript:S2583)
        ? [absenceClass, 'extra-class']
        : [],
     tooltip: toolTipText
  })
}
```
When ternary expression is evaluated, `tooltipText` and `absenceClass` will be both non-empty strings, hence the `absenceClass` condition of the ternary expression will always evaluate to true.

These issues are raised by symbolic execution analysis: in SonarJS it is applied only for function scope (not on global scope). This is why you get the issues only inside `forEach` loop as you create an arrow function there.

Hope this helps to clarify.
